### DISCUSSION TITLE
2024-04-19 - Design Patterns - Chewing the Fat 📈

### DISCUSSION THREAD

**Discussion by sethdaily** [SSWConsulting/SSW.Rules.Content#8379] [Chewing the Fat]
Hey SSWers,

1. What's your favorite design pattern?
1. Which one are you most keen to try?

Discuss below ⬇️ 

https://www.ssw.com.au/rules/common-design-patterns

**Comment by Freego1783** [comment]
Design Patterns did not really interest me in school but then I realized I was following them without really knowing it 😁

**Comment by ricksu978** [comment]
1. Observer Pattern, I love to see interactions between subscribers and publisher.
2. State Pattern.
In my past experience, the State Pattern is the least pattern I have implemented. I would love to try this pattern and apply it to a real system.

**Comment by brydeno** [comment]
1. Unit of Work because it forces you to think about what "needs" to be done for a given interaction
2. Building Repositories cleverly is typically the most interesting and challenging.

**Comment by matt-goldman** [comment]
1. Singleton 🤖 (https://chat.openai.com/share/e90433fe-8005-4ff7-b317-64034803d6c6)
2. Specification pattern - I like the concept but have not used it yet. I want to understand the practical applications better.

**Reply by AntPolkanov** [reply]
Q: Why did the singleton feel lonely in a multi-threaded environment?

A: Because it couldn't handle being shared! 😜

**Comment by tkapa** [comment]
DI is cool but I don't necessary think that it's the most important.
I think that Specification sounds kind of cool.

**Comment by AlexRothwell** [comment]
> 1. What's your favorite design pattern?

Dependency injection - it usually greatly improves the testability of a class and allows me to be more confident in its correctness 

> 2. Which one are you most keen to try?

I've not really touched the specification pattern so I'd be interested to see how that fits in

**Comment by jakebayliss** [comment]
1. Dependency Injection
2. State Pattern - Sounds like it would be fun to build

**Comment by jaydenalchin** [comment]
![image](https://github.com/SSWConsulting/SSW.Rules.Content/assets/75593567/8f04d569-e2f8-4662-a156-19b7adfa850d)

**Comment by ncn-ssw** [comment]
> 1. What's your favorite design pattern?

Spaghetti - easiest to improve on.

Actual answer is I don't have 1 - the patterns are conceptual tools, many of them attending to different problems. I try to use the pattern most applicable to the problem, with a bias towards patterns already in that code base when 2 patterns will provide similar benefits.

> 2. Which one are you most keen to try?

I find this a strange question. I've used nearly all of the patterns listed and the choice of pattern would be dependent on the problem rather than the pattern I wanted to implement.

**Comment by ozdemir-mehmet-cer** [comment]
Would be good to devote an SSW training course to patterns and how they're applied to our reference solutions

**Comment by leahy268** [comment]
Almost fell asleep watching the video.  Maybe it all made sense if you knew all those design patterns already but if you didn't then either the rule or the video - otherwise I found it a pretty boring one.

**Comment by stephanfakossw** [comment]
Does not make much sense to, but I'm not a dev :-)

**Comment by alexleessw** [comment]
NA - I'm not a developer
Have an awesome day everyone 😊

**Comment by Calinator444** [comment]
1. Depencency Injection - It improves code maintainability to abstract away the implementation details of a function. For instance if I have `checkout` class with a `payment_handler` class injected I no longer need to be concerned with how the `payment_handler` makes a payment, I only need to be concerned with it's output. This also makes it easier to unit test `checkout` because I've already seperated out all of the payment logic.

2. Mediator patten - I haven't really been on a project up until now that was large enough to dictate how classes communicate with one another.

**Comment by cvionnet** [comment]
1. Dependency Injection
2. Specification Pattern - I just heard about it a few weeks ago, thanks to @william-liebenberg

**Comment by AntPolkanov** [comment]
1. Dependency Injection - makes the code much more readable and easier to maintain.
2. Observer - we've recently started using Reactive Extensions in SSW Rewards, which allowed us to remove heaps of code and fix half a dozen bugs.

**Comment by vladislav-kir** [comment]
1. Dependency Injection
2. Repository

**Comment by TobyChurches** [comment]
1. DI
2. State - I've used this before in non-web Uni projects and it was interesting to implement

**Comment by steven0x51** [comment]
5. Repository Pattern

**Comment by PennyWalker** [comment]
I had trouble understanding this one. It was well broken down, but without the background info it was hard for me to follow and keep up. I valuable one for the devs though.

**Comment by bettybondoc** [comment]
I'm in a bit of a pickle because while I really enjoy UI/UX design, watching videos about coding always tempts me to switch back to programming. As a dev, I'm somewhere between junior and mid-level so it makes sense that I've mostly only used DI, MVC, and MVP (and spaghetti 🍝). I hadn't realised there were so many other design patterns out there. Now, I'm really starting to get curious! 🧐

**Comment by JeanThirion** [comment]
I don't have a favourite Design Pattern, nor do I think one is more important than another. It'd be like asking which one of the four wheels on a car is more important 🤪

If I REALLY had to choose - Observer is the one I find the most elegant both theoretically and in modern implementations. 

---

This is probably one of the most important topics of all time in programming. Patterns are absolutely key to successful software development. 

To Juniors - don't worry too much about learning these theoretically, just know they exist, and know that most problems have elegant solutions already.
Then with experience you'll just learn what to use and when, and you won't even think "oh, I'm using XYZ Design Pattern".

**Comment by chrisschultzssw** [comment]
Mostly went over my head... but now I want spaghetti for lunch 🤔

**Comment by SamBWagner** [comment]
Best one is the strategy pattern. It's the most generally useful in the most general places. 

I would love to find a great use case for the mediator pattern. Just installing MediatR into your project isn't understanding Mediator haha

**Comment by jimmy1992abc** [comment]
I am not a developer, I cannot understand the topic, but I am impressed.

**Comment by YalalCakirSSW** [comment]
1. I haven't deal with that much different design patterns to have a favorite one. 

2. State - sounds interesting

**Comment by Hona** [comment]
> What's your favorite design pattern?

Strategy pattern is the basis of many things like IoC & others.

>Which one are you most keen to try?

More UI design patterns, e.g. mvvm, mvc, mvp. Gain a deeper understanding of them.

**Comment by awaltos** [comment]
Repository and State patterns sound interesting, and I'd like to try those.

**Comment by amankumarrr** [comment]
- Observer Pattern: In many scenarios, we need multiple components to communicate with each other. The Observer pattern simplifies achieving this.
- State Pattern: This sounds interesting.

**Comment by MattParkerDev** [comment]
> What design pattern are you most keen to try?

the next one that is necessary to solve a problem most efficiently. I don't think we should go out with a pattern in mind looking for problems. If all you have is a hammer, everything looks like a nail.

**Comment by jeoffreyfischer** [comment]
> 1. What's your favorite design pattern?
Dependency injection

> 2. Which one are you most keen to try?
MVC

Design patterns are one of the pillars of writing codes. I have realised that I only knew a couple of them and am keen to learn more about the others. It will obviously take more than a 15min CTF to understand them so I will read about 1-2 of them everyday.

**Comment by bahjatalaadelssw** [comment]
I did follow any of that, but structured designs based on patterns is important for troubleshooting for devs

**Comment by synnoveisabel** [comment]
1. DI and Factory
2. I don't have a specific pattern as it depends on the problem you're solving

**Comment by Dhruv-0987** [comment]
Favourite: Inversion of control, game changing principal for writing loosely coupled code.

But you really have to dive deep into kind of project it is to decide which pattern is suitable.

**Comment by tombui99** [comment]
> What's your favorite design pattern?

Dependency Injection - A bit subjective as this is the pattern I spent the most time with (Angular) but it greatly helps with decoupling components, improving code maintainability and testability.

> Which one are you most keen to try?

I have heard a lot about the Mediator design pattern through Clean Architecture Workshop but only tried it once during a short client engagement. I would love to spend more time on it as it seems to work extremely well with data-heavy applications.

**Comment by AttackOnMorty** [comment]
> 1. What's your favorite design pattern?

State - The key to writing good programs is to separate [Logic and Control](https://www.doc.ic.ac.uk/~rak/papers/algorithm%20=%20logic%20+%20control.pdf). State machines can achieve this in a counterintuitive way.

**Comment by Ben0189** [comment]
Favorite design pattern - observer pattern 
Most keen to try - state pattern

**Comment by tomek-i** [comment]
Having a favourite pattern can very easily become itself an anti-pattern as you're likely to use it everywhere all the time.
I'm sure we all been there and or seen code where certain design patterns were overused and or wrongly used.

I remember in my early years when I came across a new design pattern I started using it everywhere all the time - making simple solutions a complex nightmare ...

>1. What's your favorite design pattern?

The right design pattern at the right time for the right job

>2. Which one are you most keen to try?

There are a few which I haven't encountered in the wild 
or I'm not sure of yet of their proper use cases like visitor pattern, flyweight

**Comment by drwharris** [comment]
> What's your favourite design pattern?

No favourite, it would be like picking your favourite child

> Which one are you most keen to try?

Probably Observer

**Comment by wicksipedia** [comment]
> 1. What's your favorite design pattern?

The one that helps keep my code consistent, predictable, and easy to maintain.

> 2. Which one are you most keen to try?

This feels like a loaded question - I want to try a new pattern when I need to solve a particular problem that it tackles.

If I were to choose, then Flyweight. Since it has a cool name, but I'm not sure which project would need it

**Comment by JackDevAU** [comment]
> 1. What's your favorite design pattern?

Builder! can't hate a builder pattern

> 2. Which one are you most keen to try?

Prototype

**Comment by NadeekaJK** [comment]
1.  My fav depends on the project, scope and what you're trying to do. If I'm to pick one that would be State Pattern - there are so many instances where this could clean up a full-on spaghetti code.
2. Observer Pattern - I've used it before but feels like I haven't seen its full power yet.

**Comment by duanxinhuan** [comment]
DI is the only pattern that we need to master.
All the other pattern is dust

**Comment by Harry-Ross** [comment]
I've used most of these, as I had a uni course on design patterns where we went in depth with all of the design patterns on refactoring.guru. I'm keen to try the repository pattern - seems like an interesting concept. I think a lot of them they are super important to know - as they can improve code quality and solve problems in a uniform manner across codebases i.e. reduces spaghetti 🍝.

**Reply by tomek-i** [reply]
If you worked with Entity Framework you worked with a repository pattern :-D

**Comment by christoment** [comment]
My favourite one is DI - Keen to see more State Pattern, and how to compromise it with the performance impact on a large scale :)

**Comment by arig2024** [comment]
Not a dev sounds interesting... until the monotone puts me to sleep. Im sure someone can fix that.

**Comment by jernejk** [comment]
> 1. What's your favorite design pattern?
Fluent API where you allow developers to create a chain of actions. LINQ, the way we configure .NET applications, EF Core, etc.

A well-designed, fluent API can make builder patterns, factories, queries, DIs, etc., look much more readable and maintainable.

```C#
var obj = new BuilderThing()
    .AddConfig(...)
    .RegisterXY(...)
    .Create()
    .Run();
```

This pattern can make complex code look easy to use.

> 2. Which one are you most keen to try?
The query specification pattern to abstract LINQ queries as request objects.

Although, @wicksipedia mention of Flyweight does make me curious.

**Comment by jerryluossw** [comment]
Favorite design pattern - Factory pattern, this is the first pattern I know of
Most keen to try - Observer Pattern

**Comment by Marxoz** [comment]
1. Factory Pattern. I like to hide the chaos behind the scenes.
2. Specification. Have not used it before.

**Comment by jimmidier** [comment]
1. Inversion of Control - The cornerstone of modular and multi-layered projects
2. Facade - Haven't tried before

**Comment by SofieHong** [comment]
N/A - I'm not a developer 🤣

**Comment by Aibono1225** [comment]
> What's your favorite design pattern?

DI

> Which one are you most keen to try?

Mediator

**Comment by tino-liu** [comment]
> What's your favorite design pattern?

Dependency Injection

> Which one are you most keen to try?

Facade

**Comment by babakamyljanovssw** [comment]
1. Dependency Injection and Observer
2. Specification - havent't heard or used this pattern before, sounds interesting

**Comment by bradystroud** [comment]
> 1. What's your favorite design pattern?

No favorite - I use them case by case

> 2. Which one are you most keen to try?

None in particular. The next one I use will be the next one I think I need

**Comment by danielmackay** [comment]
> 1. What's your favorite design pattern?

Inversion of Control is simple, yet incredibly powerful.  Also really like the MediatR pattern.

> 2. Which one are you most keen to try?

I've used many of these patterns, but the State pattern seems like the most interesting one that I haven't tried yet.

**Comment by MarcusIrmscherSSW** [comment]
This is way over my head, but do see how important using a Design Pattern can be.

**Comment by ozair-ashfaque** [comment]
> 1. What's your favorite design pattern?

Dependency Injection

> 2. Which one are you most keen to try?

State Pattern

**Comment by PothieuG** [comment]
What's your favorite design pattern?
-> Dependency Injection
